Small tools for genetics research 
========
Here I provide 'production' and development versions of the code I use in my research.
I hope they are of use in your research!  
If you have suggestions or improvements let me know.  
Tools are developed according to the [Small tools Manifesto](https://github.com/pjotrp/bioinformatics) principles.

---
### Measure 'special' sequences and features in WGS experiments
1. [Count_Telomeric_Sequence_Reads.py](https://github.com/jdeligt/Genetics/blob/master/Count_Telomeric_Sequence_Reads.py):  
   Extracts reads containing telomeric sequences from BAM files using [SAMBAMBA](http://lomereiter.github.io/sambamba/).
   Number of telomeric 6-mers that need to matched can be set.
   Uses multiprocessing to distribute view/count commands over different processes.
   Stores matching telomeric reads in a SAM file for further processing.  
   Uses total read count of the SAM and BAM files to generate a normalised telomere fraction.  
  
---
### WGS annotation tools
1. [Annotate_CADD_Scores_In_VCF.py](https://github.com/jdeligt/Genetics/blob/master/Annotate_CADD_Scores_In_VCF.py):  
   Annotates variants in a VCF file with [CADD scores](http://cadd.gs.washington.edu/score) using the [pre-computed](http://cadd.gs.washington.edu/download) files provided by the University of Washington.  
   Uses multiprocessing to distribute tabix lookup commands over different processes.  
  
---  
### Analyse and compare OpenArray and WGS genotyping
1. [Convert_OpenArray_CSV_to_VCF.py](https://github.com/jdeligt/Genetics/blob/master/Convert_OpenArray_CSV_to_VCF.py):
   Convert the CSV files generated by the Thermo Fisher software from an 'Open Array' genotyping experiment to VCF. This is done to facilitate comparison with WGS data. 

More scripts to extract genotypes from WGS data and compare the data are coming.

---
### Reconstruct mutational profile based on known mutational processes
1. [Mutation_Profile_Reconstructer.R](https://github.com/jdeligt/Genetics/blob/master/Mutation_Profile_Reconstructer.R):
   Based on the known mutational processes, describe by [COSMIC]( http://cancer.sanger.ac.uk/cosmic/signatures) you can determine which contributed to the signal observed in your sample. Currently the script ingests files in the ‘96 count matrix’ format. Which can be generated through NMF analysis of a VCF file, please see the [Mutational Patterns](https://github.com/CuppenResearch/MutationalPatterns) from our research group for details. The script plots the signatures used for reconstruction and their estimated contribution (based on GLMs) and provides an overview of the contributions of the signatures across the analysed samples.

---


